lib/pups/config.rb

```diff
@@ -17,44 +41,46 @@ def self.load_config(config)
     new YAML.load(config)
   end
 
-  def initialize(config)
-    @config = config
-    validate!(@config)
-
-    # Processing of the environment variables occurs first. This merges environment
-    # from the yaml templates and process ENV, and templates any variables found
-    # either via yaml or ENV.
-    @config["env"]&.each {|k,v| ENV[k] = v.to_s}
-    @config["env_template"] ||= {}
-
-    # Merging env_template variables from ENV and templates.
-    ENV.each do |k,v|
+  def self.prepare_env_template_vars(env_template, env)
+    # Merge env_template variables from env and templates.
+    env.each do |k, v|
       if k.include?('env_template_')
         key = k.gsub('env_template_', '')
-        @config["env_template"][key] = v
+        env_template[key] = v.to_s
       end
     end
+  end
 
-    # Now transform any templated environment variables prior to copying to params.
+  def self.transform_config_with_templated_vars(env_template, to_transform)
+    # Transform any templated variables prior to copying to params.
     # This has no effect if no env_template was provided.
-    @config["env_template"]&.each do |k,v|
-      ENV.each do |key, val|
-        if val.include?("{{#{k}}}")
-          ENV[key] = val.gsub("{{#{k}}}", v.to_s)
+    env_template.each do |k, v|
+      to_transform.each do |key, val|
+        if val.to_s.include?("{{#{k}}}")
+          to_transform[key] = val.gsub("{{#{k}}}", v.to_s)
         end
       end
     end
+  end
 
-    @params = @config["params"]
-    @params ||= {}
-    ENV.each do |k,v|
-      @params["$ENV_#{k}"] = v
-    end
-    inject_hooks
+  def self.combine_template_and_process_env(config, env)
+    # Merge all template env variables and process env variables, so that env
+    # variables can be provided both by configuration and runtime variables.
+    config["env"].each {|k,v| env[k] = v.to_s}
+  end
+
+  def generate_docker_run_arguments
+    output = []
+    output << Pups::Docker.generate_env_arguments(config['env'])
+    output << Pups::Docker.generate_link_arguments(config['links'])
+    output << Pups::Docker.generate_expose_arguments(config['expose'])
+    output << Pups::Docker.generate_volume_arguments(config['volumes'])
+    output << Pups::Docker.generate_label_arguments(config['labels'])
+    output.sort!.join(" ").strip
   end
 
-  def validate!(conf)
-    # raise proper errors if nodes are missing etc
+  def validate!(config)
+    # TODO: raise proper errors if nodes are missing etc
```

Comment:
hmmm, I think now is the time to implement this, either that or we remove it for now.